// sanity-template/dist/manifest/v2/common.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
exports.__esModule = true;
exports.sanityCorsOrigin = exports.tokenSpec = exports.templateMedia = exports.templateImageMedia = exports.templateTechnology = void 0;
var z = __importStar(require("zod"));
exports.templateTechnology = z.object({ id: z.string(), name: z.string(), url: z.string() });
exports.templateImageMedia = z.object({
    type: z.literal('image'),
    alt: z.string(),
    src: z.string()
});
// just keep as an alias for now
exports.templateMedia = exports.templateImageMedia;
exports.tokenSpec = z.object({
    role: z.union([z.literal('deploy-studio'), z.literal('read'), z.literal('write')]),
    label: z.string()
});
exports.sanityCorsOrigin = z.object({ origin: z.string(), allowCredentials: z.boolean() });
