// sanity-template/dist/manifest/v2/manifest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
exports.__esModule = true;
exports.templateManifest = exports.supportedDeploymentProvider = void 0;
var z = __importStar(require("zod"));
var netlify_1 = require("./netlify");
var vercel_1 = require("./vercel");
var common_1 = require("./common");
exports.supportedDeploymentProvider = z.union([netlify_1.netlifyDeployment, vercel_1.vercelDeployment]);
exports.templateManifest = z.object({
    version: z.literal(2),
    title: z.string(),
    description: z.string(),
    // Optional url to a live demo
    demoUrl: z.string().optional(),
    previewMedia: common_1.templateMedia.optional(),
    deployment: exports.supportedDeploymentProvider.optional(),
    technologies: z.array(common_1.templateTechnology).optional()
});
