// sanity-template/dist/manifest/v2/netlify.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
exports.__esModule = true;
exports.netlifyDeployment = exports.providerRequirement = void 0;
var z = __importStar(require("zod"));
var common_1 = require("./common");
exports.providerRequirement = z.literal('build-hook'); // Add future requirement flags here
var netlifySite = z.object({
    id: z.string(),
    type: z.union([z.literal('studio'), z.literal('web')]),
    title: z.string(),
    description: z.string().optional(),
    dir: z.string(),
    buildSettings: z
        .object({
        base: z.string().optional(),
        dir: z.string().optional(),
        cmd: z.string().optional()
    })
        .optional(),
    requiredTokens: z.array(common_1.tokenSpec).optional(),
    requirements: z.array(exports.providerRequirement).optional(),
    previewMedia: common_1.templateMedia.optional(),
    requiredCorsOrigins: z.array(common_1.sanityCorsOrigin).optional()
});
exports.netlifyDeployment = z.object({
    provider: z.literal('netlify'),
    sites: z.array(netlifySite)
});
