// @commercelayer/js-sdk/dist/js-sdk.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var activeResource = require('active-resource');
var _ = _interopDefault(require('lodash'));

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;

  _setPrototypeOf(subClass, superClass);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

var cleanUrl = function cleanUrl(url) {
  var lastSlash = url.lastIndexOf('/') + 1;

  if (lastSlash === url.length) {
    return url.substring(0, url.length - 1);
  }

  return url;
};
var parserParams = function parserParams(params) {
  if (!_.isEmpty(params.filter) && !_.has(params.filter, 'q')) {
    params.filter = {
      q: params.filter
    };
  }

  if (!_.isEmpty(params.fields) && !_.isEmpty(params.include)) {
    var expInclude = params.include.split(',');

    var includeKey = _.findKey(params.fields, function (v) {
      return !_.isEmpty(params.include) && v.search(params.include) !== -1 ? true : false;
    });

    var expFields = params.fields[includeKey].split(',');
    var fields = expFields.map(function (field) {
      return expInclude.includes(field) ? _.first(field.split('.')) : field;
    }).join(',');
    params.fields[includeKey] = fields;
  }

  return params;
};

var subdomain = 'yourdomain';
var library = /*#__PURE__*/activeResource.createResourceLibrary("https://" + subdomain + ".commercelayer.io/api/");
library['customRequests'] = {};

var ExtendLibrary = /*#__PURE__*/function (_library$Base) {
  _inheritsLoose(ExtendLibrary, _library$Base);

  function ExtendLibrary() {
    var _this;

    _this = _library$Base.call(this) || this;
    return _assertThisInitialized(_this) || _assertThisInitialized(_this);
  }

  ExtendLibrary.includes = function includes() {
    var _library$Base$include;

    // @ts-ignore
    this.includeMetaInfo(this["interface"]().axios.interceptors); // @ts-ignore

    for (var _len = arguments.length, params = new Array(_len), _key = 0; _key < _len; _key++) {
      params[_key] = arguments[_key];
    }

    return (_library$Base$include = _library$Base.includes).call.apply(_library$Base$include, [this].concat(params));
  };

  ExtendLibrary.select = function select() {
    var _library$Base$select;

    // @ts-ignore
    this.includeMetaInfo(this["interface"]().axios.interceptors); // @ts-ignore

    for (var _len2 = arguments.length, params = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      params[_key2] = arguments[_key2];
    }

    return (_library$Base$select = _library$Base.select).call.apply(_library$Base$select, [this].concat(params));
  };

  ExtendLibrary.order = function order(params) {
    // @ts-ignore
    this.includeMetaInfo(this["interface"]().axios.interceptors); // @ts-ignore

    return _library$Base.order.call(this, params);
  };

  ExtendLibrary.page = function page(value) {
    // @ts-ignore
    this.includeMetaInfo(this["interface"]().axios.interceptors); // @ts-ignore

    return _library$Base.page.call(this, value);
  };

  ExtendLibrary.perPage = function perPage(value) {
    // @ts-ignore
    this.includeMetaInfo(this["interface"]().axios.interceptors); // @ts-ignore

    return _library$Base.perPage.call(this, value);
  };

  ExtendLibrary.limit = function limit(value) {
    // @ts-ignore
    this.includeMetaInfo(this["interface"]().axios.interceptors); // @ts-ignore

    return _library$Base.limit.call(this, value);
  };

  ExtendLibrary.offset = function offset(value) {
    // @ts-ignore
    this.includeMetaInfo(this["interface"]().axios.interceptors); // @ts-ignore

    return _library$Base.offset.call(this, value);
  };

  ExtendLibrary.where = function where(options) {
    // @ts-ignore
    var relation = _library$Base.where.call(this, {
      q: _extends({}, options)
    });

    this.includeMetaInfo(relation["interface"]().axios.interceptors);
    return relation;
  };

  ExtendLibrary.all = function all() {
    // @ts-ignore
    this.includeMetaInfo(this["interface"]().axios.interceptors); // @ts-ignore

    return _library$Base.all.call(this);
  };

  ExtendLibrary.findBy = function findBy(options) {
    var eqOptions = {};
    Object.keys(options).map(function (v) {
      eqOptions[v + "_eq"] = options[v];
    });
    return this.where(eqOptions).first();
  };

  ExtendLibrary.withCredentials = function withCredentials(_ref) {
    var accessToken = _ref.accessToken,
        endpoint = _ref.endpoint;

    if (!this.accessToken && !this.endpoint) {
      var _this$resourceLibrary;

      if (!((_this$resourceLibrary = this.resourceLibrary.headers) != null && _this$resourceLibrary.Authorization)) {
        this.accessToken = "Bearer ";
        this.endpoint = this.resourceLibrary.baseUrl;
      } else {
        this.accessToken = this.resourceLibrary.headers.Authorization;
        this.endpoint = this.resourceLibrary.baseUrl;
      }
    }

    this.resourceLibrary.headers = {
      Authorization: "Bearer " + accessToken
    };
    this.__links = {
      related: endpoint + "/api/" + this.queryName
    };
    this.singleRequest = true;
    return this;
  };

  ExtendLibrary.includeMetaInfo = function includeMetaInfo(interceptors, klass) {
    if (interceptors === void 0) {
      interceptors = null;
    }

    if (klass === void 0) {
      klass = this;
    }

    // @ts-ignore
    var i = interceptors || this.resourceLibrary["interface"].axios.interceptors;
    this.setInterceptors(i, klass);
    return this;
  };

  ExtendLibrary.find = function find(paramKey) {
    this.includeMetaInfo();
    return _library$Base.find.call(this, paramKey);
  };

  ExtendLibrary.setCustomInterceptors = function setCustomInterceptors(interceptors) {
    library.customInterceptors = interceptors;
    return this;
  };

  ExtendLibrary.setInterceptors = function setInterceptors(interceptors, _klass) {

    var classThis = this;
    var interceptReq = interceptors == null ? void 0 : interceptors.request;
    var interceptResp = interceptors == null ? void 0 : interceptors.response;
    var respHandlers = interceptResp == null ? void 0 : interceptResp.handlers;
    interceptReq.use(function (config) {
      var _library$customInterc, _library$customInterc2;

      if (!_.isEmpty(config.params)) {
        config.params = parserParams(config.params);
      }

      if (!_.isEmpty(library == null ? void 0 : (_library$customInterc = library.customInterceptors) == null ? void 0 : _library$customInterc.request) && _.has(library == null ? void 0 : (_library$customInterc2 = library.customInterceptors) == null ? void 0 : _library$customInterc2.request, 'before')) {
        return library.customInterceptors.request.before(config);
      }

      return config;
    }, function (error) {
      var _library$customInterc3, _library$customInterc4;

      if (!_.isEmpty(library == null ? void 0 : (_library$customInterc3 = library.customInterceptors) == null ? void 0 : _library$customInterc3.request) && _.has(library == null ? void 0 : (_library$customInterc4 = library.customInterceptors) == null ? void 0 : _library$customInterc4.request, 'error')) {
        return library.customInterceptors.request.error(error);
      }

      return Promise.reject(error);
    });

    if ((respHandlers == null ? void 0 : respHandlers.length) === 1) {
      interceptResp.handlers.shift();
      interceptResp.use(function (config) {
        var _library$customInterc5, _library$customInterc6;

        if (!config['data']) return config;
        var meta = _.isArray(config.data.data) ? config.data.meta : config.data.data.meta;
        var metaCamelCase = library["interface"].toCamelCase(meta);
        classThis.meta = metaCamelCase; // config.data.collectionParent = classThis

        var url = cleanUrl(config.config.url);
        library.customRequests = {};
        library.customRequests[url] = _extends({}, config.data, {
          headers: config.headers,
          collectionParent: classThis
        });

        if (!_.isEmpty(library == null ? void 0 : (_library$customInterc5 = library.customInterceptors) == null ? void 0 : _library$customInterc5.response) && _.has(library == null ? void 0 : (_library$customInterc6 = library.customInterceptors) == null ? void 0 : _library$customInterc6.response, 'before')) {
          return library.customInterceptors.response.before(config);
        }

        return config;
      }, function (error) {
        var _library$customInterc7, _library$customInterc8;

        if (!_.isEmpty(library == null ? void 0 : (_library$customInterc7 = library.customInterceptors) == null ? void 0 : _library$customInterc7.response) && _.has(library == null ? void 0 : (_library$customInterc8 = library.customInterceptors) == null ? void 0 : _library$customInterc8.response, 'error')) {
          return library.customInterceptors.response.error(error);
        }

        return Promise.reject(error);
      });
    }
  };

  return ExtendLibrary;
}(library.Base);

ExtendLibrary.accessToken = '';
ExtendLibrary.endpoint = '';
ExtendLibrary.singleRequest = false;
ExtendLibrary.meta = {};

var setMetaByRequest = function setMetaByRequest(child) {
  _.map(library.customRequests, function (req) {
    if (!_.isEmpty(req.headers)) {
      child.setHeaders(req.headers);
    }

    if (_.isArray(req.data)) {
      // @ts-ignore
      var childData = _.first(req.data.filter(function (v) {
        return child.id === v.id;
      }));

      if (!_.isEmpty(childData)) {
        var collectionMeta = library["interface"].toCamelCase(req.meta);
        var meta = library["interface"].toCamelCase(childData.meta);
        child.setMetaInfo(meta);
        child.setCollectionMetaInfo(collectionMeta);
      }
    } else {
      var _meta = library["interface"].toCamelCase(req.data.meta);

      child.setMetaInfo(_meta);
    }
  });
};

ExtendLibrary.afterBuild(function () {
  // @ts-ignore
  delete this.__meta; // @ts-ignore

  delete this.__collectionMeta; // @ts-ignore

  delete this.__headers;
});
ExtendLibrary.afterRequest(function () {

  setMetaByRequest(this); // @ts-ignore

  if (this.constructor.singleRequest) {
    // @ts-ignore
    this.constructor.resourceLibrary.baseUrl = this.constructor.endpoint; // @ts-ignore

    this.constructor.resourceLibrary.headers = {
      // @ts-ignore
      Authorization: this.constructor.accessToken
    }; // @ts-ignore

    this.constructor.__links = null; // @ts-ignore

    this.constructor.singleRequest = false;
  }
});

activeResource.CollectionResponse.prototype.getMetaInfo = function () {
  var firstItem = this.first();

  if (_.isEmpty(firstItem)) {
    return {};
  }

  return _.has(firstItem, '__collectionMeta') ? firstItem.__collectionMeta : firstItem.getMetaInfo();
};

activeResource.CollectionResponse.prototype.getHeaders = function () {
  var firstItem = this.first();

  if (_.isEmpty(firstItem)) {
    return {};
  }

  return _.has(firstItem, '__headers') ? firstItem.__headers : firstItem.getHeaders();
};

activeResource.CollectionResponse.prototype.pageCount = function () {
  var firstItem = this.first();

  if (_.isEmpty(firstItem)) {
    return 0;
  }

  return _.has(firstItem, '__collectionMeta.pageCount') ? firstItem.__collectionMeta.pageCount : this.first().pageCount();
};

activeResource.CollectionResponse.prototype.recordCount = function () {
  var firstItem = this.first();

  if (_.isEmpty(firstItem)) {
    return 0;
  }

  return _.has(firstItem, '__collectionMeta.recordCount') ? firstItem.__collectionMeta.recordCount : this.first().recordCount();
};

activeResource.CollectionResponse.prototype.withCredentials = function (_ref2) {
  var accessToken = _ref2.accessToken;
  library.headers = {
    Authorization: "Bearer " + accessToken
  };
  library.singleRequest = true;
  return this;
};

library.Base = ExtendLibrary;

var BaseClass = /*#__PURE__*/function (_library$Base) {
  _inheritsLoose(BaseClass, _library$Base);

  function BaseClass() {
    var _this;

    _this = _library$Base.apply(this, arguments) || this;
    _this.__meta = {};
    _this.__collectionMeta = {};
    _this.__headers = {};
    return _this;
  } // errors: () => Errors<this>
  // update: (attrs: object, callback?: (response: this) => void) => this


  var _proto = BaseClass.prototype;

  _proto.getMetaInfo = function getMetaInfo() {
    return this.__meta;
  };

  _proto.getHeaders = function getHeaders() {
    return this.__headers;
  };

  _proto.pageCount = function pageCount() {
    return this.__meta.pageCount;
  };

  _proto.recordCount = function recordCount() {
    return this.__meta.recordCount;
  };

  _proto.mode = function mode() {
    return this.__meta.mode;
  };

  _proto.setMetaInfo = function setMetaInfo(meta) {
    return this.__meta = meta;
  };

  _proto.setHeaders = function setHeaders(headers) {
    return this.__headers = headers;
  };

  _proto.setCollectionMetaInfo = function setCollectionMetaInfo(meta) {
    return this.__collectionMeta = meta;
  };

  _proto.klass = function klass() {
    // @ts-ignore
    var interceptors = _library$Base.prototype.klass.call(this)["interface"]().axios.interceptors; // @ts-ignore


    this.constructor.includeMetaInfo(interceptors); // @ts-ignore

    return _library$Base.prototype.klass.call(this);
  };

  _proto.association = function association(name) {
    // @ts-ignore
    var interceptors = _library$Base.prototype.association.call(this, name)["interface"]().axios.interceptors; // @ts-ignore


    this.constructor.includeMetaInfo(interceptors); // @ts-ignore

    return _library$Base.prototype.association.call(this, name);
  };

  _proto.withCredentials = function withCredentials(_ref) {
    var accessToken = _ref.accessToken,
        endpoint = _ref.endpoint;
    // @ts-ignore
    this.constructor.includeMetaInfo(); // @ts-ignore

    if (!this.constructor.accessToken && !this.constructor.endpoint) {
      var _this$constructor$res;

      // @ts-ignore
      if (!((_this$constructor$res = this.constructor.resourceLibrary.headers) != null && _this$constructor$res.Authorization)) {
        // @ts-ignore
        this.constructor.accessToken = "Bearer "; // @ts-ignore

        this.constructor.endpoint = this.constructor.resourceLibrary.baseUrl;
      } else {
        // @ts-ignore
        this.constructor.accessToken = this.constructor.resourceLibrary.headers.Authorization; // @ts-ignore

        this.constructor.endpoint = this.constructor.resourceLibrary.baseUrl;
      }
    } // @ts-ignore


    this.constructor.resourceLibrary.headers = {
      Authorization: "Bearer " + accessToken
    }; // @ts-ignore

    this.__links = {
      // @ts-ignore
      related: endpoint + "/api/" + this.constructor.queryName,
      // @ts-ignore
      self: endpoint + "/api/" + this.constructor.queryName + "/" + this.id
    }; // @ts-ignore

    this.constructor.singleRequest = true;
    return this;
  };

  _proto.setCustomInterceptors = function setCustomInterceptors(interceptors) {
    // @ts-ignore
    this.constructor.customInterceptors = interceptors;
    return this;
  };

  return BaseClass;
}(library.Base);

var AddressCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(AddressCollection, _BaseClass);

  function AddressCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  AddressCollection.define = function define() {
    this.attributes('business', 'firstName', 'lastName', 'company', 'fullName', 'line1', 'line2', 'city', 'zipCode', 'stateCode', 'countryCode', 'phone', 'fullAddress', 'name', 'email', 'notes', 'lat', 'lng', 'isLocalized', 'isGeocoded', 'providerName', 'mapUrl', 'staticMapUrl', 'billingInfo', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('geocoder', {
      className: 'Geocoder'
    });
  };

  return AddressCollection;
}(BaseClass);
AddressCollection.className = 'Address';
var Address = /*#__PURE__*/library.createResource(AddressCollection);

var AttachmentCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(AttachmentCollection, _BaseClass);

  function AttachmentCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  AttachmentCollection.define = function define() {
    this.attributes('name', 'description', 'url', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.belongsTo('attachable', {
      polymorphic: true
    });
  };

  return AttachmentCollection;
}(BaseClass);
AttachmentCollection.className = 'Attachment';
var Attachment = /*#__PURE__*/library.createResource(AttachmentCollection);
Attachment.afterBuild(function () {
  // @ts-ignore
  if (this.attachableId) delete this.attachableId; // @ts-ignore

  if (this.attachableType) delete this.attachableType;
});

var CustomerAddressCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CustomerAddressCollection, _BaseClass);

  function CustomerAddressCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CustomerAddressCollection.define = function define() {
    this.attributes('name', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('customer', {
      className: 'Customer'
    });
    this.hasOne('address', {
      className: 'Address'
    });
  };

  return CustomerAddressCollection;
}(BaseClass);
CustomerAddressCollection.className = 'CustomerAddress';
var CustomerAddress = /*#__PURE__*/library.createResource(CustomerAddressCollection);

var CustomerGroupCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CustomerGroupCollection, _BaseClass);

  function CustomerGroupCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CustomerGroupCollection.define = function define() {
    this.attributes('name', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('priceList', {
      className: 'PriceList'
    });
    this.hasMany('customers', {
      className: 'Customer'
    });
  };

  return CustomerGroupCollection;
}(BaseClass);
CustomerGroupCollection.className = 'CustomerGroup';
var CustomerGroup = /*#__PURE__*/library.createResource(CustomerGroupCollection);

var CustomerPasswordResetCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CustomerPasswordResetCollection, _BaseClass);

  function CustomerPasswordResetCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CustomerPasswordResetCollection.define = function define() {
    this.attributes('customerEmail', 'resetPasswordToken', 'customerPassword', '_resetPasswordToken', 'resetPasswordAt', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('customer', {
      className: 'Customer'
    });
  };

  return CustomerPasswordResetCollection;
}(BaseClass);
CustomerPasswordResetCollection.className = 'CustomerPasswordReset';
var CustomerPasswordReset = /*#__PURE__*/library.createResource(CustomerPasswordResetCollection);

var CustomerPaymentSourceCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CustomerPaymentSourceCollection, _BaseClass);

  function CustomerPaymentSourceCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CustomerPaymentSourceCollection.define = function define() {
    this.attributes('name', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('customer', {
      className: 'Customer'
    });
    this.belongsTo('paymentSource', {
      polymorphic: true
    });
  };

  return CustomerPaymentSourceCollection;
}(BaseClass);
CustomerPaymentSourceCollection.className = 'CustomerPaymentSource';
var CustomerPaymentSource = /*#__PURE__*/library.createResource(CustomerPaymentSourceCollection);
CustomerPaymentSource.afterBuild(function () {
  // @ts-ignore
  if (this.paymentSourceId) delete this.paymentSourceId; // @ts-ignore

  if (this.paymentSourceType) delete this.paymentSourceType;
});

var CustomerSubscriptionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CustomerSubscriptionCollection, _BaseClass);

  function CustomerSubscriptionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CustomerSubscriptionCollection.define = function define() {
    this.attributes('customerEmail', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('customer', {
      className: 'Customer'
    });
  };

  return CustomerSubscriptionCollection;
}(BaseClass);
CustomerSubscriptionCollection.className = 'CustomerSubscription';
var CustomerSubscription = /*#__PURE__*/library.createResource(CustomerSubscriptionCollection);

var CustomerCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CustomerCollection, _BaseClass);

  function CustomerCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CustomerCollection.define = function define() {
    this.attributes('email', 'hasPassword', 'password', 'passwordBlank', 'passwordPresent', 'status', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('customerGroup', {
      className: 'CustomerGroup'
    });
    this.hasMany('customerAddresses', {
      className: 'CustomerAddress'
    });
    this.hasMany('customerPaymentSources', {
      className: 'CustomerPaymentSource'
    });
    this.hasMany('customerSubscriptions', {
      className: 'CustomerSubscription'
    });
    this.hasMany('orders', {
      className: 'Order'
    });
  };

  return CustomerCollection;
}(BaseClass);
CustomerCollection.className = 'Customer';
var Customer = /*#__PURE__*/library.createResource(CustomerCollection);

var DeliveryLeadTimeCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(DeliveryLeadTimeCollection, _BaseClass);

  function DeliveryLeadTimeCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  DeliveryLeadTimeCollection.define = function define() {
    this.attributes('minHours', 'maxHours', 'minDays', 'maxDays', 'id', 'createdAt', 'updatedAt', 'referenceOrigin', 'metadata');
    this.hasOne('stockLocation', {
      className: 'StockLocation'
    });
    this.hasOne('shippingMethod', {
      className: 'ShippingMethod'
    });
  };

  return DeliveryLeadTimeCollection;
}(BaseClass);
DeliveryLeadTimeCollection.className = 'DeliveryLeadTime';
var DeliveryLeadTime = /*#__PURE__*/library.createResource(DeliveryLeadTimeCollection);

var ImportCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ImportCollection, _BaseClass);

  function ImportCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ImportCollection.define = function define() {
    this.attributes('resourceType', 'parentResourceId', 'status', 'startedAt', 'completedAt', 'inputs', 'errorsCount', 'warningsCount', 'destroyedCount', 'errorsLog', 'warningsLog', 'cleanupRecords', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
  };

  return ImportCollection;
}(BaseClass);
ImportCollection.className = 'Import';
var Import = /*#__PURE__*/library.createResource(ImportCollection);

var InStockSubscriptionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(InStockSubscriptionCollection, _BaseClass);

  function InStockSubscriptionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  InStockSubscriptionCollection.define = function define() {
    this.attributes('status', 'customerEmail', 'skuCode', '_activate', '_deactivate', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('customer', {
      className: 'Customer'
    });
    this.hasOne('sku', {
      className: 'Sku'
    });
  };

  return InStockSubscriptionCollection;
}(BaseClass);
InStockSubscriptionCollection.className = 'InStockSubscription';
var InStockSubscription = /*#__PURE__*/library.createResource(InStockSubscriptionCollection);

var InventoryModelCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(InventoryModelCollection, _BaseClass);

  function InventoryModelCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  InventoryModelCollection.define = function define() {
    this.attributes('name', 'id', 'stockLocationsCutoff', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasMany('inventoryStockLocations', {
      className: 'InventoryStockLocation'
    });
  };

  return InventoryModelCollection;
}(BaseClass);
InventoryModelCollection.className = 'InventoryModel';
var InventoryModel = /*#__PURE__*/library.createResource(InventoryModelCollection);

var LineItemOptionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(LineItemOptionCollection, _BaseClass);

  function LineItemOptionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  LineItemOptionCollection.define = function define() {
    this.attributes('name', 'quantity', 'currencyCode', 'unitAmountCents', 'unitAmountFloat', 'formattedUnitAmount', 'totalAmountCents', 'totalAmountFloat', 'formattedTotalAmount', 'delayHours', 'delayDays', 'options', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('lineItem', {
      className: 'LineItem'
    });
    this.hasOne('skuOption', {
      className: 'SkuOption'
    });
  };

  return LineItemOptionCollection;
}(BaseClass);
LineItemOptionCollection.className = 'LineItemOption';
var LineItemOption = /*#__PURE__*/library.createResource(LineItemOptionCollection);

var LineItemCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(LineItemCollection, _BaseClass);

  function LineItemCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  LineItemCollection.define = function define() {
    this.attributes('skuCode', 'quantity', '_updateQuantity', 'currencyCode', 'unitAmountCents', 'unitAmountFloat', 'formattedUnitAmount', 'optionsAmountCents', 'optionsAmountFloat', 'formattedOptionsAmount', 'totalAmountCents', 'totalAmountFloat', 'formattedTotalAmount', 'name', 'imageUrl', 'taxRate', 'taxBreakdown', 'itemType', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
    this.belongsTo('item', {
      polymorphic: true
    });
    this.hasMany('lineItemOptions', {
      className: 'LineItemOption'
    });
  };

  return LineItemCollection;
}(BaseClass);
LineItemCollection.className = 'LineItem';
var LineItem = /*#__PURE__*/library.createResource(LineItemCollection);
LineItem.afterBuild(function () {
  // @ts-ignore
  if (this.itemId) delete this.itemId; // @ts-ignore

  if (this.itemType) delete this.itemType;
});

var MarketCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(MarketCollection, _BaseClass);

  function MarketCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  MarketCollection.define = function define() {
    this.attributes('number', 'name', 'facebookPixelId', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('merchant', {
      className: 'Merchant'
    });
    this.hasOne('priceList', {
      className: 'PriceList'
    });
    this.hasOne('inventoryModel', {
      className: 'InventoryModel'
    });
    this.hasMany('customerGroup', {
      className: 'CustomerGroup'
    });
  };

  return MarketCollection;
}(BaseClass);
MarketCollection.className = 'Market';
var Market = /*#__PURE__*/library.createResource(MarketCollection);

var MerchantCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(MerchantCollection, _BaseClass);

  function MerchantCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  MerchantCollection.define = function define() {
    this.attributes('name', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('address', {
      className: 'Address'
    });
  };

  return MerchantCollection;
}(BaseClass);
MerchantCollection.className = 'Merchant';
var Merchant = /*#__PURE__*/library.createResource(MerchantCollection);

var OrderCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(OrderCollection, _BaseClass);

  function OrderCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  OrderCollection.define = function define() {
    this.attributes('number', 'status', 'paymentStatus', 'fulfillmentStatus', 'guest', 'editable', 'placeable', 'customerEmail', 'customerPassword', 'languageCode', 'currencyCode', 'taxIncluded', 'taxRate', 'freightTaxable', 'countryCode', 'shippingCountryCodeLock', 'couponCode', 'subtotalAmountCents', 'subtotalAmountFloat', 'formattedSubtotalAmount', 'shippingAmountCents', 'shippingAmountFloat', 'formattedShippingAmount', 'paymentMethodAmountCents', 'paymentMethodAmountFloat', 'formattedPaymentMethodAmount', 'discountAmountCents', 'discountAmountFloat', 'formattedDiscountAmount', 'totalTaxAmountCents', 'totalTaxAmountFloat', 'formattedTotalTaxAmount', 'subtotalTaxAmountCents', 'subtotalTaxAmountFloat', 'formattedSubtotalTaxAmount', 'shippingTaxAmountCents', 'shippingTaxAmountFloat', 'formattedShippingTaxAmount', 'paymentMethodTaxAmountCents', 'paymentMethodTaxAmountFloat', 'formattedPaymentMethodTaxAmount', 'discountTaxAmountCents', 'discountTaxAmountFloat', 'formattedDiscountTaxAmount', 'totalAmountCents', 'totalAmountFloat', 'formattedTotalAmount', 'totalTaxableAmountCents', 'totalTaxableAmountFloat', 'formattedTotalTaxableAmount', 'subtotalTaxableAmountCents', 'subtotalTaxableAmountFloat', 'formattedSubtotalTaxableAmount', 'shippingTaxableAmountCents', 'shippingTaxableAmountFloat', 'formattedShippingTaxableAmount', 'paymentMethodTaxableAmountCents', 'paymentMethodTaxableAmountFloat', 'formattedPaymentMethodTaxableAmount', 'discountTaxableAmountCents', 'discountTaxableAmountFloat', 'formattedDiscountTaxableAmount', 'totalAmountWithTaxesCents', 'totalAmountWithTaxesFloat', 'formattedTotalAmountWithTaxes', 'feesAmountCents', 'feesAmountFloat', 'formattedFeesAmount', 'skusCount', 'lineItemOptionsCount', 'shipmentsCount', 'paymentSourceDetails', 'token', 'cartUrl', 'returnUrl', 'termsUrl', 'privacyUrl', 'checkoutUrl', '_place', '_cancel', '_approve', '_capture', '_refund', '_updateTaxes', '_billingAddressCloneId', '_shippingAddressCloneId', '_customerPaymentSourceId', '_shippingAddressSameAsBilling', '_billingAddressSameAsShipping', '_savePaymentSourceToCustomerWallet', '_saveShippingAddressToCustomerAddressBook', '_saveBillingAddressToCustomerAddressBook', '_refresh', 'placedAt', 'approvedAt', 'cancelledAt', 'paymentUpdatedAt', 'fulfillmentUpdatedAt', 'id', 'archivedAt', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    }); // Required

    this.hasOne('customer', {
      className: 'Customer'
    });
    this.hasOne('shippingAddress', {
      className: 'Address'
    });
    this.hasOne('billingAddress', {
      className: 'Address'
    });
    this.hasOne('paymentMethod', {
      className: 'PaymentMethod'
    });
    this.hasMany('availablePaymentMethods', {
      className: 'PaymentMethod'
    });
    this.hasMany('lineItems', {
      className: 'LineItem'
    });
    this.hasMany('shipments', {
      className: 'Shipment'
    });
    this.belongsTo('paymentSource', {
      polymorphic: true
    });
  };

  return OrderCollection;
}(BaseClass);
OrderCollection.className = 'Order';
var Order = /*#__PURE__*/library.createResource(OrderCollection);
Order.afterBuild(function () {
  // @ts-ignore
  if (this.paymentSourceId) delete this.paymentSourceId; // @ts-ignore

  if (this.paymentSourceType) delete this.paymentSourceType;
});

var ParcelCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ParcelCollection, _BaseClass);

  function ParcelCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ParcelCollection.define = function define() {
    this.attributes('number', 'weight', 'unitOfWeight', 'eelPfc', 'contentsType', 'contentsExplanation', 'customsCertify', 'customsSigner', 'nonDeliveryOption', 'restrictionType', 'restrictionComments', 'customsInfoRequired', 'shippingLabelUrl', 'shippingLabelFileType', 'shippingLabelSize', 'shippingLabelResolution', 'trackingNumber', 'trackingStatus', 'trackingStatusDetail', 'trackingStatusUpdatedAt', 'trackingDetails', 'carrierWeightOz', 'signedBy', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('shipment', {
      className: 'Shipment'
    });
    this.hasMany('parcelLineItems', {
      className: 'LineItem'
    });
  };

  return ParcelCollection;
}(BaseClass);
ParcelCollection.className = 'Parcel';
var Parcel = /*#__PURE__*/library.createResource(ParcelCollection);

var PaymentMethodCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PaymentMethodCollection, _BaseClass);

  function PaymentMethodCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PaymentMethodCollection.define = function define() {
    this.attributes('paymentSourceType', 'name', 'disabledAt', 'priceAmountCents', 'priceAmountFloat', 'formattedPriceAmount', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('paymentGateway', {
      className: 'PaymentGateway'
    });
  };

  return PaymentMethodCollection;
}(BaseClass);
PaymentMethodCollection.className = 'PaymentMethod';
var PaymentMethod = /*#__PURE__*/library.createResource(PaymentMethodCollection);

var PaymentGatewayCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PaymentGatewayCollection, _BaseClass);

  function PaymentGatewayCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PaymentGatewayCollection.define = function define() {
    this.attributes('id');
  };

  return PaymentGatewayCollection;
}(BaseClass);
PaymentGatewayCollection.className = 'PaymentGateway';
var PaymentGateway = /*#__PURE__*/library.createResource(PaymentGatewayCollection);

var PaypalPaymentCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PaypalPaymentCollection, _BaseClass);

  function PaypalPaymentCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PaypalPaymentCollection.define = function define() {
    this.attributes('returnUrl', 'cancelUrl', 'noteToPayer', 'paypalPayerId', 'name', 'paypalId', 'status', 'approvalUrl', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
  };

  return PaypalPaymentCollection;
}(BaseClass);
PaypalPaymentCollection.className = 'PaypalPayment';
var PaypalPayment = /*#__PURE__*/library.createResource(PaypalPaymentCollection);

var PriceCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PriceCollection, _BaseClass);

  function PriceCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PriceCollection.define = function define() {
    this.attributes('currencyCode', 'skuCode', 'amountCents', 'amountFloat', 'formattedAmount', 'compareAtAmountCents', 'compareAtAmountFloat', 'formattedCompareAtAmount', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('priceList', {
      className: 'PriceList'
    });
    this.hasOne('sku', {
      className: 'Sku'
    });
  };

  return PriceCollection;
}(BaseClass);
PriceCollection.className = 'Price';
var Price = /*#__PURE__*/library.createResource(PriceCollection);

var PriceListCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PriceListCollection, _BaseClass);

  function PriceListCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PriceListCollection.define = function define() {
    this.attributes('name', 'currencyCode', 'taxIncluded', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasMany('prices', {
      className: 'Price'
    });
  };

  return PriceListCollection;
}(BaseClass);
PriceListCollection.className = 'PriceList';
var PriceList = /*#__PURE__*/library.createResource(PriceListCollection);

var ShipmentCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ShipmentCollection, _BaseClass);

  function ShipmentCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ShipmentCollection.define = function define() {
    this.attributes('number', 'status', 'currencyCode', 'costAmountCents', 'costAmountFloat', 'formattedCostAmount', '_onHold', '_picking', '_packing', '_readyToShip', '_ship', '_getRates', 'selectedRateId', '_purchase', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('shippingCategory', {
      className: 'ShippingCategory'
    });
    this.hasOne('stockLocation', {
      className: 'StockLocation'
    });
    this.hasOne('shippingAddress', {
      className: 'Address'
    });
    this.hasOne('shippingMethod', {
      className: 'ShippingMethod'
    });
    this.hasMany('shipmentLineItems', {
      className: 'ShipmentLineItem'
    });
    this.hasMany('availableShippingMethods', {
      className: 'ShippingMethod'
    });
    this.hasMany('parcels', {
      className: 'Parcel'
    });
    this.hasMany('attachments', {
      className: 'Attachment'
    });
    this.hasMany('stockTransfers', {
      className: 'StockTransfer'
    });
  };

  return ShipmentCollection;
}(BaseClass);
ShipmentCollection.className = 'Shipment';
var Shipment = /*#__PURE__*/library.createResource(ShipmentCollection);

var ShippingCategoryCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ShippingCategoryCollection, _BaseClass);

  function ShippingCategoryCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ShippingCategoryCollection.define = function define() {
    this.attributes('name', 'id', 'createdAt', 'updatedAt', 'referenceOrigin', 'metadata');
    this.hasMany('skus', {
      className: 'Sku'
    });
  };

  return ShippingCategoryCollection;
}(BaseClass);
ShippingCategoryCollection.className = 'ShippingCategory';
var ShippingCategory = /*#__PURE__*/library.createResource(ShippingCategoryCollection);

var ShippingMethodCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ShippingMethodCollection, _BaseClass);

  function ShippingMethodCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ShippingMethodCollection.define = function define() {
    this.attributes('name', 'disabledAt', 'currencyCode', 'priceAmountCents', 'priceAmountFloat', 'formattedPriceAmount', 'freeOverAmountCents', 'freeOverAmountFloat', 'formattedFreeOverAmount', 'priceAmountForShipmentCents', 'priceAmountForShipmentFloat', 'formattedPriceAmountForShipment', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('shippingZone', {
      className: 'ShippingZone'
    });
    this.hasOne('shippingCategory', {
      className: 'ShippingCategory'
    });
    this.hasOne('deliveryLeadTimeForShipment', {
      className: 'DeliveryLeadTime'
    });
  };

  return ShippingMethodCollection;
}(BaseClass);
ShippingMethodCollection.className = 'ShippingMethod';
var ShippingMethod = /*#__PURE__*/library.createResource(ShippingMethodCollection);

var ShippingZoneCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ShippingZoneCollection, _BaseClass);

  function ShippingZoneCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ShippingZoneCollection.define = function define() {
    this.attributes('name', 'countryCodeRegex', 'notCountryCodeRegex', 'stateCodeRegex', 'notStateCodeRegex', 'zipCodeRegex', 'notZipCodeRegex', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
  };

  return ShippingZoneCollection;
}(BaseClass);
ShippingZoneCollection.className = 'ShippingZone';
var ShippingZone = /*#__PURE__*/library.createResource(ShippingZoneCollection);

var SkuOptionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(SkuOptionCollection, _BaseClass);

  function SkuOptionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  SkuOptionCollection.define = function define() {
    this.attributes('name', 'description', 'priceAmountCents', 'priceAmountFloat', 'formattedPriceAmount', 'delayHours', 'delayDays', 'skuCodeRegex', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
  };

  return SkuOptionCollection;
}(BaseClass);
SkuOptionCollection.className = 'SkuOption';
var SkuOption = /*#__PURE__*/library.createResource(SkuOptionCollection);

var SkuCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(SkuCollection, _BaseClass);

  function SkuCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  SkuCollection.define = function define() {
    this.attributes('code', 'name', 'description', 'imageUrl', 'tagNames', 'piecesPerPack', 'weight', 'unitOfWeight', 'inventory', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('shippingCategory', {
      className: 'ShippingCategory'
    });
    this.hasMany('prices', {
      className: 'Price'
    });
    this.hasMany('stockItems', {
      className: 'StockItem'
    });
    this.hasMany('deliveryLeadTimes', {
      className: 'DeliveryLeadTime'
    });
    this.hasMany('skuOptions', {
      className: 'SkuOption'
    });
    this.hasMany('availablePromotions', {
      className: 'Promotion'
    });
  };

  return SkuCollection;
}(BaseClass);
SkuCollection.className = 'Sku';
var Sku = /*#__PURE__*/library.createResource(SkuCollection);

var SkuListCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(SkuListCollection, _BaseClass);

  function SkuListCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  SkuListCollection.define = function define() {
    this.attributes('name', 'slug', 'description', 'manual', 'skuCodeRegex', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasMany('skus', {
      className: 'Sku'
    });
    this.hasMany('skuListItems', {
      className: 'SkuListItem'
    });
  };

  return SkuListCollection;
}(BaseClass);
SkuListCollection.className = 'SkuList';
var SkuList = /*#__PURE__*/library.createResource(SkuListCollection);

var SkuListItemCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(SkuListItemCollection, _BaseClass);

  function SkuListItemCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  SkuListItemCollection.define = function define() {
    this.attributes('position', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('sku', {
      className: 'Sku'
    });
    this.hasOne('skuList', {
      className: 'SkuList'
    });
  };

  return SkuListItemCollection;
}(BaseClass);
SkuListItemCollection.className = 'SkuListItem';
var SkuListItem = /*#__PURE__*/library.createResource(SkuListItemCollection);

var StockItemCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(StockItemCollection, _BaseClass);

  function StockItemCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  StockItemCollection.define = function define() {
    this.attributes('skuCode', 'quantity', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('stockLocation', {
      className: 'StockLocation'
    });
    this.hasOne('sku', {
      className: 'Sku'
    });
  };

  return StockItemCollection;
}(BaseClass);
StockItemCollection.className = 'StockItem';
var StockItem = /*#__PURE__*/library.createResource(StockItemCollection);

var InventoryStockLocationCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(InventoryStockLocationCollection, _BaseClass);

  function InventoryStockLocationCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  InventoryStockLocationCollection.define = function define() {
    this.attributes('priority', 'onHold', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('stockLocation', {
      className: 'StockLocation'
    });
    this.hasOne('inventoryModel', {
      className: 'InventoryModel'
    });
  };

  return InventoryStockLocationCollection;
}(BaseClass);
InventoryStockLocationCollection.className = 'InventoryStockLocation';
var InventoryStockLocation = /*#__PURE__*/library.createResource(InventoryStockLocationCollection);

var StockLocationCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(StockLocationCollection, _BaseClass);

  function StockLocationCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  StockLocationCollection.define = function define() {
    this.attributes('name', 'labelFormat', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('address', {
      className: 'Address'
    });
    this.hasMany('InventoryStockLocations', {
      className: 'InventoryStockLocation'
    });
    this.hasMany('stockItems', {
      className: 'StockItem'
    });
  };

  return StockLocationCollection;
}(BaseClass);
StockLocationCollection.className = 'StockLocation';
var StockLocation = /*#__PURE__*/library.createResource(StockLocationCollection);

var WebhookCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(WebhookCollection, _BaseClass);

  function WebhookCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  WebhookCollection.define = function define() {
    this.attributes('topic', 'callbackUrl', 'includeResources', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
  };

  return WebhookCollection;
}(BaseClass);
WebhookCollection.className = 'Webhook';
var Webhook = /*#__PURE__*/library.createResource(WebhookCollection);

var WireTransferCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(WireTransferCollection, _BaseClass);

  function WireTransferCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  WireTransferCollection.define = function define() {
    this.attributes('id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
  };

  return WireTransferCollection;
}(BaseClass);
WireTransferCollection.className = 'WireTransfer';
var WireTransfer = /*#__PURE__*/library.createResource(WireTransferCollection);

var init = function init(_ref) {
  var accessToken = _ref.accessToken,
      endpoint = _ref.endpoint,
      interceptors = _ref.interceptors;
  library.baseUrl = endpoint + "/api/";
  library.headers = {
    Authorization: "Bearer " + accessToken
  };
  library.customInterceptors = interceptors;
};

var GeocoderCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(GeocoderCollection, _BaseClass);

  function GeocoderCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  GeocoderCollection.define = function define() {
    this.attributes();
  };

  return GeocoderCollection;
}(BaseClass);
GeocoderCollection.className = 'Geocoder';
var Geocoder = /*#__PURE__*/library.createResource(GeocoderCollection);

var PaymentSourceCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PaymentSourceCollection, _BaseClass);

  function PaymentSourceCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PaymentSourceCollection.define = function define() {};

  return PaymentSourceCollection;
}(BaseClass);
PaymentSourceCollection.className = 'PaymentSource';
var PaymentSource = /*#__PURE__*/library.createResource(PaymentSourceCollection);

var ItemCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ItemCollection, _BaseClass);

  function ItemCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ItemCollection.define = function define() {
    this.attributes();
  };

  return ItemCollection;
}(BaseClass);
ItemCollection.className = 'Item';
var Item = /*#__PURE__*/library.createResource(ItemCollection);

var AttachableCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(AttachableCollection, _BaseClass);

  function AttachableCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  AttachableCollection.define = function define() {
    this.attributes();
  };

  return AttachableCollection;
}(BaseClass);
AttachableCollection.className = 'Attachable';
var Attachable = /*#__PURE__*/library.createResource(AttachableCollection);

var GiftCardCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(GiftCardCollection, _BaseClass);

  function GiftCardCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  GiftCardCollection.define = function define() {
    this.attributes('status', 'code', 'currencyCode', 'initialBalanceCents', 'initialBalanceFloat', 'formattedInitialBalance', 'balanceCents', 'balanceFloat', 'formattedBalance', 'balanceMaxCents', 'balanceMaxFloat', 'formattedBalanceMax', 'balanceLog', 'singleUse', 'rechargeable', 'imageUrl', 'expiresAt', 'referenceOrigin', 'recipientEmail', '_purchase', '_activate', '_deactivate', '_balanceChangeCents', 'id', 'createAt', 'updateAt', 'reference', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('giftCardRecipient', {
      className: 'GiftCardRecipient'
    });
  };

  return GiftCardCollection;
}(BaseClass);
GiftCardCollection.className = 'GiftCard';
var GiftCard = /*#__PURE__*/library.createResource(GiftCardCollection);

var GiftCardRecipientCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(GiftCardRecipientCollection, _BaseClass);

  function GiftCardRecipientCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  GiftCardRecipientCollection.define = function define() {
    this.attributes('email', 'firstName', 'lastName', 'referenceOrigin', 'id', 'createAt', 'updateAt', 'reference', 'metadata');
    this.hasOne('customer', {
      className: 'Customer'
    });
  };

  return GiftCardRecipientCollection;
}(BaseClass);
GiftCardRecipientCollection.className = 'GiftCardRecipient';
var GiftCardRecipient = /*#__PURE__*/library.createResource(GiftCardRecipientCollection);

var AdyenPaymentCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(AdyenPaymentCollection, _BaseClass);

  function AdyenPaymentCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  AdyenPaymentCollection.define = function define() {
    this.attributes('paymentMethod', 'paymentRequestData', 'paymentRequestDetails', 'paymentResponse', '_authorize', '_details', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
  };

  return AdyenPaymentCollection;
}(BaseClass);
AdyenPaymentCollection.className = 'AdyenPayment';
var AdyenPayment = /*#__PURE__*/library.createResource(AdyenPaymentCollection);

var BraintreePaymentCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(BraintreePaymentCollection, _BaseClass);

  function BraintreePaymentCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  BraintreePaymentCollection.define = function define() {
    this.attributes('clientToken', 'paymentMethodNonce', 'options', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
  };

  return BraintreePaymentCollection;
}(BaseClass);
BraintreePaymentCollection.className = 'BraintreePayment';
var BraintreePayment = /*#__PURE__*/library.createResource(BraintreePaymentCollection);

var StripePaymentCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(StripePaymentCollection, _BaseClass);

  function StripePaymentCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  StripePaymentCollection.define = function define() {
    this.attributes('clientSecret', 'options', 'id', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
  };

  return StripePaymentCollection;
}(BaseClass);
StripePaymentCollection.className = 'StripePayment';
var StripePayment = /*#__PURE__*/library.createResource(StripePaymentCollection);

var ExternalPaymentCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ExternalPaymentCollection, _BaseClass);

  function ExternalPaymentCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ExternalPaymentCollection.define = function define() {
    this.attributes('createdAt', 'id', 'metadata', 'options', 'paymentSourceToken', 'reference', 'referenceOrigin', 'updatedAt');
    this.hasOne('order', {
      className: 'Order'
    });
  };

  return ExternalPaymentCollection;
}(BaseClass);
ExternalPaymentCollection.className = 'ExternalPayment';
var ExternalPayment = /*#__PURE__*/library.createResource(ExternalPaymentCollection);

var PackageCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PackageCollection, _BaseClass);

  function PackageCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PackageCollection.define = function define() {
    this.attributes('code', 'createdAt', 'length', 'metadata', 'name', 'reference', 'referenceOrigin', 'unitOfLength', 'updatedAt', 'width');
    this.hasOne('stockLocation', {
      className: 'StockLocation'
    });
    this.hasMany('parcels', {
      className: 'Parcel'
    });
  };

  return PackageCollection;
}(BaseClass);
PackageCollection.className = 'Package';
var Package = /*#__PURE__*/library.createResource(PackageCollection);

var AdjustmentCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(AdjustmentCollection, _BaseClass);

  function AdjustmentCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  AdjustmentCollection.define = function define() {
    this.attributes('name', 'currencyCode', 'amountCents', 'amountFloat', 'formattedAmount', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
  };

  return AdjustmentCollection;
}(BaseClass);
AdjustmentCollection.className = 'Adjustment';
var Adjustment = /*#__PURE__*/library.createResource(AdjustmentCollection);

var CaptureCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CaptureCollection, _BaseClass);

  function CaptureCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CaptureCollection.define = function define() {
    this.attributes('number', 'currencyCode', 'amountCents', 'amountFloat', 'formattedAmount', 'succeeded', 'message', 'errorCode', 'errorDetail', 'token', 'gatewayTransactionId', '_refund', '_refundAmountCents', 'refundAmountCents', 'refundAmountFloat', 'formattedRefundAmount', 'refundBalanceCents', 'refundBalanceFloat', 'formattedRefundBalance', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
    this.hasOne('referenceAuthorization', {
      className: 'Authorization'
    });
    this.hasMany('refunds', {
      className: 'Refund'
    });
  };

  return CaptureCollection;
}(BaseClass);
CaptureCollection.className = 'Capture';
var Capture = /*#__PURE__*/library.createResource(CaptureCollection);

var AuthorizationCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(AuthorizationCollection, _BaseClass);

  function AuthorizationCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  AuthorizationCollection.define = function define() {
    this.attributes('number', 'currencyCode', 'amountCents', 'amountFloat', 'formattedAmount', 'succeeded', 'message', 'errorCode', 'errorDetail', 'cvvCode', 'cvvMessage', 'avsCode', 'avsMessage', 'fraudReview', 'gatewayTransactionId', '_capture', '_captureAmountCents', 'captureAmountCents', 'captureAmountFloat', 'formattedCaptureAmount', 'captureBalanceCents', 'captureBalanceFloat', 'formattedCaptureBalance', '_void', 'voidBalanceCents', 'voidBalanceFloat', 'formattedVoidBalance', 'name', 'currencyCode', 'amountCents', 'amountFloat', 'formattedAmount', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
    this.hasMany('captures', {
      className: 'Capture'
    });
    this.hasMany('voids', {
      className: 'Void'
    });
  };

  return AuthorizationCollection;
}(BaseClass);
AuthorizationCollection.className = 'Authorization';
var Authorization = /*#__PURE__*/library.createResource(AuthorizationCollection);

var BillingInfoValidationRuleCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(BillingInfoValidationRuleCollection, _BaseClass);

  function BillingInfoValidationRuleCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  BillingInfoValidationRuleCollection.define = function define() {
    this.attributes('createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
  };

  return BillingInfoValidationRuleCollection;
}(BaseClass);
BillingInfoValidationRuleCollection.className = 'BillingInfoValidationRule';
var BillingInfoValidationRule = /*#__PURE__*/library.createResource(BillingInfoValidationRuleCollection);

var RefundCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(RefundCollection, _BaseClass);

  function RefundCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  RefundCollection.define = function define() {
    this.attributes('number', 'currencyCode', 'amountCents', 'amountFloat', 'formattedAmount', 'succeeded', 'message', 'errorCode', 'errorDetail', 'token', 'gatewayTransactionId', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
    this.hasOne('referenceCapture', {
      className: 'Capture'
    });
  };

  return RefundCollection;
}(BaseClass);
RefundCollection.className = 'Refund';
var Refund = /*#__PURE__*/library.createResource(RefundCollection);

var VoidCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(VoidCollection, _BaseClass);

  function VoidCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  VoidCollection.define = function define() {
    this.attributes('number', 'currencyCode', 'amountCents', 'amountFloat', 'formattedAmount', 'succeeded', 'message', 'errorCode', 'errorDetail', 'token', 'gatewayTransactionId', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
    this.hasOne('referenceAuthorization', {
      className: 'Authorization'
    });
  };

  return VoidCollection;
}(BaseClass);
VoidCollection.className = 'Void';
var Void = /*#__PURE__*/library.createResource(VoidCollection);

var CouponCodesPromotionRuleCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CouponCodesPromotionRuleCollection, _BaseClass);

  function CouponCodesPromotionRuleCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CouponCodesPromotionRuleCollection.define = function define() {
    this.attributes('createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('promotion', {
      className: 'Promotion'
    });
  };

  return CouponCodesPromotionRuleCollection;
}(BaseClass);
CouponCodesPromotionRuleCollection.className = 'CouponCodesPromotionRule';
var CouponCodesPromotionRule = /*#__PURE__*/library.createResource(CouponCodesPromotionRuleCollection);

var CouponCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(CouponCollection, _BaseClass);

  function CouponCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  CouponCollection.define = function define() {
    this.attributes('code', 'usageLimit', 'usageCount', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('promotionRule', {
      className: 'PromotionRule'
    });
  };

  return CouponCollection;
}(BaseClass);
CouponCollection.className = 'Coupon';
var Coupon = /*#__PURE__*/library.createResource(CouponCollection);

var ExternalPromotionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ExternalPromotionCollection, _BaseClass);

  function ExternalPromotionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ExternalPromotionCollection.define = function define() {
    this.attributes('name', 'startsAt', 'expiresAt', 'totalUsageLimit', 'totalUsageCount', 'active', 'promotionUrl', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('orderAmountPromotionRule', {
      className: 'OrderAmountPromotionRule'
    });
    this.hasOne('skuListPromotionRule', {
      className: 'SkuListPromotionRule'
    });
    this.hasOne('couponCodesPromotionRule', {
      className: 'CouponCodesPromotionRule'
    });
    this.hasMany('promotionRules', {
      className: 'PromotionRule'
    });
  };

  return ExternalPromotionCollection;
}(BaseClass);
ExternalPromotionCollection.className = 'ExternalPromotion';
var ExternalPromotion = /*#__PURE__*/library.createResource(ExternalPromotionCollection);

var FixedAmountPromotionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(FixedAmountPromotionCollection, _BaseClass);

  function FixedAmountPromotionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  FixedAmountPromotionCollection.define = function define() {
    this.attributes('name', 'startsAt', 'expiresAt', 'totalUsageLimit', 'totalUsageCount', 'active', 'fixedAmountCents', 'fixedAmountFloat', 'formattedFixedAmount', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('orderAmountPromotionRule', {
      className: 'OrderAmountPromotionRule'
    });
    this.hasOne('skuListPromotionRule', {
      className: 'SkuListPromotionRule'
    });
    this.hasOne('couponCodesPromotionRule', {
      className: 'CouponCodesPromotionRule'
    });
    this.hasMany('promotionRules', {
      className: 'PromotionRule'
    });
  };

  return FixedAmountPromotionCollection;
}(BaseClass);
FixedAmountPromotionCollection.className = 'FixedAmountPromotion';
var FixedAmountPromotion = /*#__PURE__*/library.createResource(FixedAmountPromotionCollection);

var FreeShippingPromotionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(FreeShippingPromotionCollection, _BaseClass);

  function FreeShippingPromotionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  FreeShippingPromotionCollection.define = function define() {
    this.attributes('name', 'startsAt', 'expiresAt', 'totalUsageLimit', 'totalUsageCount', 'active', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('orderAmountPromotionRule', {
      className: 'OrderAmountPromotionRule'
    });
    this.hasOne('skuListPromotionRule', {
      className: 'SkuListPromotionRule'
    });
    this.hasOne('couponCodesPromotionRule', {
      className: 'CouponCodesPromotionRule'
    });
    this.hasMany('promotionRules', {
      className: 'PromotionRule'
    });
  };

  return FreeShippingPromotionCollection;
}(BaseClass);
FreeShippingPromotionCollection.className = 'FreeShippingPromotion';
var FreeShippingPromotion = /*#__PURE__*/library.createResource(FreeShippingPromotionCollection);

var OrderAmountPromotionRuleCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(OrderAmountPromotionRuleCollection, _BaseClass);

  function OrderAmountPromotionRuleCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  OrderAmountPromotionRuleCollection.define = function define() {
    this.attributes('orderAmountCents', 'orderAmountFloat', 'formattedOrderAmount', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('promotion', {
      className: 'Promotion'
    });
  };

  return OrderAmountPromotionRuleCollection;
}(BaseClass);
OrderAmountPromotionRuleCollection.className = 'OrderAmountPromotionRule';
var OrderAmountPromotionRule = /*#__PURE__*/library.createResource(OrderAmountPromotionRuleCollection);

var PercentageDiscountPromotionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PercentageDiscountPromotionCollection, _BaseClass);

  function PercentageDiscountPromotionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PercentageDiscountPromotionCollection.define = function define() {
    this.attributes('name', 'startsAt', 'expiresAt', 'totalUsageLimit', 'totalUsageCount', 'active', 'percentage', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('orderAmountPromotionRule', {
      className: 'OrderAmountPromotionRule'
    });
    this.hasOne('skuListPromotionRule', {
      className: 'SkuListPromotionRule'
    });
    this.hasOne('skuList', {
      className: 'SkuList'
    });
    this.hasOne('couponCodesPromotionRule', {
      className: 'CouponCodesPromotionRule'
    });
    this.hasMany('promotionRules', {
      className: 'PromotionRule'
    });
    this.hasMany('skus', {
      className: 'Sku'
    });
  };

  return PercentageDiscountPromotionCollection;
}(BaseClass);
PercentageDiscountPromotionCollection.className = 'PercentageDiscountPromotion';
var PercentageDiscountPromotion = /*#__PURE__*/library.createResource(PercentageDiscountPromotionCollection);

var PromotionRuleCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PromotionRuleCollection, _BaseClass);

  function PromotionRuleCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PromotionRuleCollection.define = function define() {
    this.attributes('createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('promotion', {
      className: 'Promotion'
    });
  };

  return PromotionRuleCollection;
}(BaseClass);
PromotionRuleCollection.className = 'PromotionRule';
var PromotionRule = /*#__PURE__*/library.createResource(PromotionRuleCollection);

var PromotionCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(PromotionCollection, _BaseClass);

  function PromotionCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  PromotionCollection.define = function define() {
    this.attributes('name', 'startsAt', 'expiresAt', 'totalUsageLimit', 'totalUsageCount', 'active', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('market', {
      className: 'Market'
    });
    this.hasOne('orderAmountPromotionRule', {
      className: 'OrderAmountPromotionRule'
    });
    this.hasOne('skuListPromotionRule', {
      className: 'SkuListPromotionRule'
    });
    this.hasOne('couponCodesPromotionRule', {
      className: 'CouponCodesPromotionRule'
    });
    this.hasMany('promotionRules', {
      className: 'PromotionRule'
    });
  };

  return PromotionCollection;
}(BaseClass);
PromotionCollection.className = 'Promotion';
var Promotion = /*#__PURE__*/library.createResource(PromotionCollection);

var SkuListPromotionRuleCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(SkuListPromotionRuleCollection, _BaseClass);

  function SkuListPromotionRuleCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  SkuListPromotionRuleCollection.define = function define() {
    this.attributes('createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('promotion', {
      className: 'Promotion'
    });
    this.hasOne('skuList', {
      className: 'SkuList'
    });
    this.hasMany('skus', {
      className: 'Sku'
    });
  };

  return SkuListPromotionRuleCollection;
}(BaseClass);
SkuListPromotionRuleCollection.className = 'SkuListPromotionRule';
var SkuListPromotionRule = /*#__PURE__*/library.createResource(SkuListPromotionRuleCollection);

var InventoryReturnLocationCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(InventoryReturnLocationCollection, _BaseClass);

  function InventoryReturnLocationCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  InventoryReturnLocationCollection.define = function define() {
    this.attributes('priority', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('stockLocation', {
      className: 'StockLocation'
    });
    this.hasOne('inventoryModel', {
      className: 'InventoryModel'
    });
  };

  return InventoryReturnLocationCollection;
}(BaseClass);
InventoryReturnLocationCollection.className = 'InventoryReturnLocation';
var InventoryModel$1 = /*#__PURE__*/library.createResource(InventoryReturnLocationCollection);

var ReturnCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ReturnCollection, _BaseClass);

  function ReturnCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ReturnCollection.define = function define() {
    this.attributes('number', 'customerEmail', 'status', 'skusCount', 'approvedAt', 'rejectedAt', 'shippedAt', 'receivedAt', 'archivedAt', '_request', '_approve', '_reject', '_ship', '_receive', '_restock', '_archive', '_unarchive', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('order', {
      className: 'Order'
    });
    this.hasOne('customer', {
      className: 'Customer'
    });
    this.hasOne('inventoryReturnLocation', {
      className: 'InventoryReturnLocation'
    });
    this.hasOne('stockLocation', {
      className: 'StockLocation'
    });
    this.hasOne('originAddress', {
      className: 'Address'
    });
    this.hasOne('destinationAddress', {
      className: 'Address'
    });
    this.hasMany('returnLineItems', {
      className: 'ReturnLineItem'
    });
    this.hasMany('attachments', {
      className: 'Attachment'
    });
  };

  return ReturnCollection;
}(BaseClass);
ReturnCollection.className = 'Return';
var InventoryModel$2 = /*#__PURE__*/library.createResource(ReturnCollection);

var ExternalTaxCalculatorCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ExternalTaxCalculatorCollection, _BaseClass);

  function ExternalTaxCalculatorCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ExternalTaxCalculatorCollection.define = function define() {
    this.attributes('createdAt', 'id', 'metadata', 'name', 'reference', 'referenceOrigin', 'updatedAt', 'taxCalculatorUrl');
    this.hasMany('taxCategories', {
      className: 'TaxCategory'
    });
  };

  return ExternalTaxCalculatorCollection;
}(BaseClass);
ExternalTaxCalculatorCollection.className = 'ExternalTaxCalculator';
var ExternalTaxCalculator = /*#__PURE__*/library.createResource(ExternalTaxCalculatorCollection);

var ManualTaxCalculatorCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ManualTaxCalculatorCollection, _BaseClass);

  function ManualTaxCalculatorCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ManualTaxCalculatorCollection.define = function define() {
    this.attributes('createdAt', 'id', 'metadata', 'name', 'reference', 'referenceOrigin', 'updatedAt');
    this.hasMany('taxCategories', {
      className: 'TaxCategory'
    });
    this.hasMany('taxRules', {
      className: 'TaxRule'
    });
  };

  return ManualTaxCalculatorCollection;
}(BaseClass);
ManualTaxCalculatorCollection.className = 'ManualTaxCalculator';
var ManualTaxCalculator = /*#__PURE__*/library.createResource(ManualTaxCalculatorCollection);

var TaxCalculatorCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(TaxCalculatorCollection, _BaseClass);

  function TaxCalculatorCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  TaxCalculatorCollection.define = function define() {
    this.attributes('createdAt', 'id', 'metadata', 'name', 'reference', 'referenceOrigin', 'updatedAt');
    this.hasMany('taxCategories', {
      className: 'TaxCategory'
    });
  };

  return TaxCalculatorCollection;
}(BaseClass);
TaxCalculatorCollection.className = 'TaxCalculator';
var TaxCalculator = /*#__PURE__*/library.createResource(TaxCalculatorCollection);

var TaxRuleCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(TaxRuleCollection, _BaseClass);

  function TaxRuleCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  TaxRuleCollection.define = function define() {
    this.attributes('adjustmentTaxable', 'breakDown', 'countryCodeRegex', 'createdAt', 'freightTaxable', 'giftCardTaxable', 'id', 'metadata', 'name', 'notCountryCodeRegex', 'notStateCodeRegex', 'notZipCodeRegex', 'paymentMethodTaxable', 'reference', 'referenceOrigin', 'stateCodeRegex', 'taxRate', 'updatedAt', 'zipCodeRegex');
    this.hasOne('manualTaxCalculator', {
      className: 'ManualTaxCalculator'
    });
  };

  return TaxRuleCollection;
}(BaseClass);
TaxRuleCollection.className = 'TaxRule';
var TaxRule = /*#__PURE__*/library.createResource(TaxRuleCollection);

var TaxjarAccountCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(TaxjarAccountCollection, _BaseClass);

  function TaxjarAccountCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  TaxjarAccountCollection.define = function define() {
    this.attributes('apiKey', 'createdAt', 'id', 'metadata', 'name', 'reference', 'referenceOrigin', 'updatedAt');
    this.hasMany('taxCategories', {
      className: 'TaxCategory'
    });
  };

  return TaxjarAccountCollection;
}(BaseClass);
TaxjarAccountCollection.className = 'TaxjarAccount';
var TaxjarAccount = /*#__PURE__*/library.createResource(TaxjarAccountCollection);

var ReturnLineItemCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ReturnLineItemCollection, _BaseClass);

  function ReturnLineItemCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ReturnLineItemCollection.define = function define() {
    this.attributes('skuCode', 'name', 'quantity', '_restock', 'returnReason', 'restockedAt', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('return', {
      className: 'Return'
    });
    this.hasOne('lineItem', {
      className: 'LineItem'
    });
  };

  return ReturnLineItemCollection;
}(BaseClass);
ReturnLineItemCollection.className = 'ReturnLineItem';
var InventoryModel$3 = /*#__PURE__*/library.createResource(ReturnLineItemCollection);

var ShipmentLineItemCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(ShipmentLineItemCollection, _BaseClass);

  function ShipmentLineItemCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  ShipmentLineItemCollection.define = function define() {
    this.attributes('skuCode', 'quantity', 'createdAt', 'updatedAt', 'reference', 'referenceOrigin', 'metadata');
    this.hasOne('stockItem', {
      className: 'StockItem'
    });
    this.hasOne('lineItem', {
      className: 'LineItem'
    });
    this.hasOne('shipment', {
      className: 'Shipment'
    });
  };

  return ShipmentLineItemCollection;
}(BaseClass);
ShipmentLineItemCollection.className = 'ShipmentLineItem';
var ShipmentLineItem = /*#__PURE__*/library.createResource(ShipmentLineItemCollection);

var TaxCategoryCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(TaxCategoryCollection, _BaseClass);

  function TaxCategoryCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  TaxCategoryCollection.define = function define() {
    this.attributes('code', 'createdAt', 'id', 'metadata', 'reference', 'referenceOrigin', 'skuCode', 'updatedAt');
    this.hasOne('sku', {
      className: 'Sku'
    });
    this.hasOne('taxCalculator', {
      className: 'TaxCalculator'
    });
  };

  return TaxCategoryCollection;
}(BaseClass);
TaxCategoryCollection.className = 'TaxCategory';
var TaxCategory = /*#__PURE__*/library.createResource(TaxCategoryCollection);

var StockTransferCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(StockTransferCollection, _BaseClass);

  function StockTransferCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  StockTransferCollection.define = function define() {
    this.attributes('_cancel', '_complete', '_inTransit', '_picking', '_upcoming', 'cancelledAt', 'completedAt', 'createdAt', 'id', 'metadata', 'quantity', 'reference', 'referenceOrigin', 'skuCode', 'status', 'updatedAt');
    this.hasOne('sku', {
      className: 'Sku'
    });
    this.hasOne('originStockLocation', {
      className: 'StockLocation'
    });
    this.hasOne('destinationStockLocation', {
      className: 'StockLocation'
    });
    this.hasOne('shipment', {
      className: 'Shipment'
    });
    this.hasOne('lineItem', {
      className: 'LineItem'
    });
  };

  return StockTransferCollection;
}(BaseClass);
StockTransferCollection.className = 'StockTransfer';
var StockTransfer = /*#__PURE__*/library.createResource(StockTransferCollection);

var AvalaraAccountCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(AvalaraAccountCollection, _BaseClass);

  function AvalaraAccountCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  AvalaraAccountCollection.define = function define() {
    this.attributes('companyCode', 'createdAt', 'createdAt', 'ddp', 'id', 'metadata', 'name', 'password', 'reference', 'referenceOrigin', 'updatedAt');
    this.hasMany('taxCategories', {
      className: 'TaxCategory'
    });
  };

  return AvalaraAccountCollection;
}(BaseClass);
AvalaraAccountCollection.className = 'AvalaraAccount';
var AvalaraAccount = /*#__PURE__*/library.createResource(AvalaraAccountCollection);

var OrganizationCollection = /*#__PURE__*/function (_BaseClass) {
  _inheritsLoose(OrganizationCollection, _BaseClass);

  function OrganizationCollection() {
    return _BaseClass.apply(this, arguments) || this;
  }

  OrganizationCollection.define = function define() {
    this.attributes('acceptanceDisabled', 'accountDisabled', 'contrastColor', 'createdAt', 'discountDisabled', 'domain', 'faviconUrl', 'gtmId', 'gtmIdTest', 'logoUrl', 'metadata', 'name', 'primaryColor', 'reference', 'referenceOrigin', 'slug', 'updatedAt');
  };

  return OrganizationCollection;
}(BaseClass);
OrganizationCollection.className = 'Organization';
OrganizationCollection.queryName = 'organization';
var Organization = /*#__PURE__*/library.createResource(OrganizationCollection);

var CLayer = {
  Address: Address,
  Adjustment: Adjustment,
  AdyenPayment: AdyenPayment,
  Attachable: Attachable,
  Attachment: Attachment,
  Authorization: Authorization,
  AvalaraAccount: AvalaraAccount,
  BillingInfoValidationRule: BillingInfoValidationRule,
  BraintreePayment: BraintreePayment,
  Capture: Capture,
  Coupon: Coupon,
  CouponCodesPromotionRule: CouponCodesPromotionRule,
  Customer: Customer,
  CustomerAddress: CustomerAddress,
  CustomerGroup: CustomerGroup,
  CustomerPasswordReset: CustomerPasswordReset,
  CustomerPaymentSource: CustomerPaymentSource,
  CustomerSubscription: CustomerSubscription,
  DeliveryLeadTime: DeliveryLeadTime,
  ExternalPayment: ExternalPayment,
  ExternalPromotion: ExternalPromotion,
  ExternalTaxCalculator: ExternalTaxCalculator,
  FixedAmountPromotion: FixedAmountPromotion,
  FreeShippingPromotion: FreeShippingPromotion,
  Geocoder: Geocoder,
  GiftCard: GiftCard,
  GiftCardRecipient: GiftCardRecipient,
  Import: Import,
  InStockSubscription: InStockSubscription,
  InventoryModel: InventoryModel,
  InventoryReturnLocation: InventoryModel$1,
  InventoryStockLocation: InventoryStockLocation,
  Item: Item,
  LineItem: LineItem,
  LineItemOption: LineItemOption,
  ManualTaxCalculator: ManualTaxCalculator,
  Market: Market,
  Merchant: Merchant,
  Order: Order,
  OrderAmountPromotionRule: OrderAmountPromotionRule,
  Organization: Organization,
  Package: Package,
  Parcel: Parcel,
  PaymentGateway: PaymentGateway,
  PaymentMethod: PaymentMethod,
  PaymentSource: PaymentSource,
  PaypalPayment: PaypalPayment,
  PercentageDiscountPromotion: PercentageDiscountPromotion,
  Price: Price,
  PriceList: PriceList,
  Promotion: Promotion,
  PromotionRule: PromotionRule,
  Refund: Refund,
  Return: InventoryModel$2,
  ReturnLineItem: InventoryModel$3,
  Shipment: Shipment,
  ShipmentLineItem: ShipmentLineItem,
  ShippingCategory: ShippingCategory,
  ShippingMethod: ShippingMethod,
  ShippingZone: ShippingZone,
  Sku: Sku,
  SkuList: SkuList,
  SkuListItem: SkuListItem,
  SkuListPromotionRule: SkuListPromotionRule,
  SkuOption: SkuOption,
  StockItem: StockItem,
  StockLocation: StockLocation,
  StockTransfer: StockTransfer,
  StripePayment: StripePayment,
  TaxCalculator: TaxCalculator,
  TaxCategory: TaxCategory,
  TaxRule: TaxRule,
  TaxjarAccount: TaxjarAccount,
  Void: Void,
  Webhook: Webhook,
  WireTransfer: WireTransfer,
  init: init
};
var initCLayer = init;

exports.Address = Address;
exports.AddressCollection = AddressCollection;
exports.Adjustment = Adjustment;
exports.AdjustmentCollection = AdjustmentCollection;
exports.AdyenPayment = AdyenPayment;
exports.AdyenPaymentCollection = AdyenPaymentCollection;
exports.Attachable = Attachable;
exports.AttachableCollection = AttachableCollection;
exports.Attachment = Attachment;
exports.AttachmentCollection = AttachmentCollection;
exports.Authorization = Authorization;
exports.AuthorizationCollection = AuthorizationCollection;
exports.AvalaraAccount = AvalaraAccount;
exports.AvalaraAccountCollection = AvalaraAccountCollection;
exports.BillingInfoValidationRule = BillingInfoValidationRule;
exports.BillingInfoValidationRuleCollection = BillingInfoValidationRuleCollection;
exports.BraintreePayment = BraintreePayment;
exports.BraintreePaymentCollection = BraintreePaymentCollection;
exports.Capture = Capture;
exports.CaptureCollection = CaptureCollection;
exports.Coupon = Coupon;
exports.CouponCodesPromotionRule = CouponCodesPromotionRule;
exports.CouponCodesPromotionRuleCollection = CouponCodesPromotionRuleCollection;
exports.CouponCollection = CouponCollection;
exports.Customer = Customer;
exports.CustomerAddress = CustomerAddress;
exports.CustomerAddressCollection = CustomerAddressCollection;
exports.CustomerCollection = CustomerCollection;
exports.CustomerGroup = CustomerGroup;
exports.CustomerGroupCollection = CustomerGroupCollection;
exports.CustomerPasswordReset = CustomerPasswordReset;
exports.CustomerPasswordResetCollection = CustomerPasswordResetCollection;
exports.CustomerPaymentSource = CustomerPaymentSource;
exports.CustomerPaymentSourceCollection = CustomerPaymentSourceCollection;
exports.CustomerSubscription = CustomerSubscription;
exports.CustomerSubscriptionCollection = CustomerSubscriptionCollection;
exports.DeliveryLeadTime = DeliveryLeadTime;
exports.DeliveryLeadTimeCollection = DeliveryLeadTimeCollection;
exports.ExternalPayment = ExternalPayment;
exports.ExternalPaymentCollection = ExternalPaymentCollection;
exports.ExternalPromotion = ExternalPromotion;
exports.ExternalPromotionCollection = ExternalPromotionCollection;
exports.ExternalTaxCalculator = ExternalTaxCalculator;
exports.ExternalTaxCalculatorCollection = ExternalTaxCalculatorCollection;
exports.FixedAmountPromotion = FixedAmountPromotion;
exports.FixedAmountPromotionCollection = FixedAmountPromotionCollection;
exports.FreeShippingPromotion = FreeShippingPromotion;
exports.FreeShippingPromotionCollection = FreeShippingPromotionCollection;
exports.Geocoder = Geocoder;
exports.GeocoderCollection = GeocoderCollection;
exports.GiftCard = GiftCard;
exports.GiftCardCollection = GiftCardCollection;
exports.GiftCardRecipient = GiftCardRecipient;
exports.GiftCardRecipientCollection = GiftCardRecipientCollection;
exports.Import = Import;
exports.ImportCollection = ImportCollection;
exports.InStockSubscription = InStockSubscription;
exports.InStockSubscriptionCollection = InStockSubscriptionCollection;
exports.InventoryModel = InventoryModel;
exports.InventoryModelCollection = InventoryModelCollection;
exports.InventoryReturnLocation = InventoryModel$1;
exports.InventoryReturnLocationCollection = InventoryReturnLocationCollection;
exports.InventoryStockLocation = InventoryStockLocation;
exports.InventoryStockLocationCollection = InventoryStockLocationCollection;
exports.Item = Item;
exports.LineItem = LineItem;
exports.LineItemCollection = LineItemCollection;
exports.LineItemOption = LineItemOption;
exports.LineItemOptionCollection = LineItemOptionCollection;
exports.ManualTaxCalculator = ManualTaxCalculator;
exports.ManualTaxCalculatorCollection = ManualTaxCalculatorCollection;
exports.Market = Market;
exports.MarketCollection = MarketCollection;
exports.Merchant = Merchant;
exports.MerchantCollection = MerchantCollection;
exports.Order = Order;
exports.OrderAmountPromotionRule = OrderAmountPromotionRule;
exports.OrderAmountPromotionRuleCollection = OrderAmountPromotionRuleCollection;
exports.OrderCollection = OrderCollection;
exports.Organization = Organization;
exports.OrganizationCollection = OrganizationCollection;
exports.Package = Package;
exports.PackageCollection = PackageCollection;
exports.Parcel = Parcel;
exports.ParcelCollection = ParcelCollection;
exports.PaymentGateway = PaymentGateway;
exports.PaymentGatewayCollection = PaymentGatewayCollection;
exports.PaymentMethod = PaymentMethod;
exports.PaymentMethodCollection = PaymentMethodCollection;
exports.PaymentSource = PaymentSource;
exports.PaymentSourceCollection = PaymentSourceCollection;
exports.PaypalPayment = PaypalPayment;
exports.PaypalPaymentCollection = PaypalPaymentCollection;
exports.PercentageDiscountPromotion = PercentageDiscountPromotion;
exports.PercentageDiscountPromotionCollection = PercentageDiscountPromotionCollection;
exports.Price = Price;
exports.PriceCollection = PriceCollection;
exports.PriceList = PriceList;
exports.PriceListCollection = PriceListCollection;
exports.Promotion = Promotion;
exports.PromotionCollection = PromotionCollection;
exports.PromotionRule = PromotionRule;
exports.PromotionRuleCollection = PromotionRuleCollection;
exports.Refund = Refund;
exports.RefundCollection = RefundCollection;
exports.Return = InventoryModel$2;
exports.ReturnCollection = ReturnCollection;
exports.ReturnLineItem = InventoryModel$3;
exports.ReturnLineItemCollection = ReturnLineItemCollection;
exports.Shipment = Shipment;
exports.ShipmentCollection = ShipmentCollection;
exports.ShipmentLineItem = ShipmentLineItem;
exports.ShipmentLineItemCollection = ShipmentLineItemCollection;
exports.ShippingCategory = ShippingCategory;
exports.ShippingCategoryCollection = ShippingCategoryCollection;
exports.ShippingMethod = ShippingMethod;
exports.ShippingMethodCollection = ShippingMethodCollection;
exports.ShippingZone = ShippingZone;
exports.ShippingZoneCollection = ShippingZoneCollection;
exports.Sku = Sku;
exports.SkuCollection = SkuCollection;
exports.SkuList = SkuList;
exports.SkuListCollection = SkuListCollection;
exports.SkuListItem = SkuListItem;
exports.SkuListItemCollection = SkuListItemCollection;
exports.SkuListPromotionRule = SkuListPromotionRule;
exports.SkuListPromotionRuleCollection = SkuListPromotionRuleCollection;
exports.SkuOption = SkuOption;
exports.SkuOptionCollection = SkuOptionCollection;
exports.StockItem = StockItem;
exports.StockItemCollection = StockItemCollection;
exports.StockLocation = StockLocation;
exports.StockLocationCollection = StockLocationCollection;
exports.StockTransfer = StockTransfer;
exports.StockTransferCollection = StockTransferCollection;
exports.StripePayment = StripePayment;
exports.StripePaymentCollection = StripePaymentCollection;
exports.TaxCalculator = TaxCalculator;
exports.TaxCalculatorCollection = TaxCalculatorCollection;
exports.TaxCategory = TaxCategory;
exports.TaxCategoryCollection = TaxCategoryCollection;
exports.TaxRule = TaxRule;
exports.TaxRuleCollection = TaxRuleCollection;
exports.TaxjarAccount = TaxjarAccount;
exports.TaxjarAccountCollection = TaxjarAccountCollection;
exports.Void = Void;
exports.VoidCollection = VoidCollection;
exports.Webhook = Webhook;
exports.WebhookCollection = WebhookCollection;
exports.WireTransfer = WireTransfer;
exports.WireTransferCollection = WireTransferCollection;
exports.default = CLayer;
exports.initCLayer = initCLayer;
