// sanity-template/dist/manifest/v1/manifest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
exports.__esModule = true;
exports.templateManifest = exports.deployment = exports.netlify = exports.sanityCorsOrigin = exports.providerRequirement = exports.tokenSpec = exports.templateMedia = exports.templateImageMedia = exports.templateTechnology = void 0;
var z = __importStar(require("zod"));
exports.templateTechnology = z.object({ id: z.string(), name: z.string(), url: z.string() });
exports.templateImageMedia = z.object({
    type: z.literal('image'),
    alt: z.string(),
    src: z.string()
});
// just keep as an alias for now
exports.templateMedia = exports.templateImageMedia;
exports.tokenSpec = z.object({
    role: z.union([z.literal('deploy-studio'), z.literal('read'), z.literal('write')]),
    label: z.string()
});
exports.providerRequirement = z.literal('build-hook'); // Add future requirement flags here
exports.sanityCorsOrigin = z.object({ origin: z.string(), allowCredentials: z.boolean() });
exports.netlify = z.object({
    name: z.literal('netlify'),
    config: z
        .object({
        base: z.string().optional(),
        dir: z.string().optional(),
        cmd: z.string().optional()
    })
        .optional(),
    requirements: z.array(exports.providerRequirement).optional()
});
exports.deployment = z.object({
    id: z.string(),
    type: z.union([z.literal('studio'), z.literal('web')]),
    title: z.string(),
    description: z.string().optional(),
    dir: z.string(),
    provider: exports.netlify,
    previewMedia: exports.templateMedia,
    requiredCorsOrigins: z.array(exports.sanityCorsOrigin).optional(),
    requiredTokens: z.array(exports.tokenSpec).optional()
});
exports.templateManifest = z.object({
    version: z.union([z.literal(1), z.literal(0)]),
    title: z.string(),
    description: z.string(),
    previewMedia: exports.templateMedia.optional(),
    deployments: z.array(exports.deployment),
    technologies: z.array(exports.templateTechnology).optional()
});
