// react-instantsearch-core/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createConnector: true,
  instantSearchContext: true,
  InstantSearchConsumer: true,
  InstantSearchProvider: true,
  HIGHLIGHT_TAGS: true,
  version: true,
  translatable: true,
  Configure: true,
  ExperimentalConfigureRelatedItems: true,
  QueryRuleContext: true,
  Index: true,
  InstantSearch: true,
  connectAutoComplete: true,
  connectBreadcrumb: true,
  connectConfigure: true,
  EXPERIMENTAL_connectConfigureRelatedItems: true,
  connectCurrentRefinements: true,
  connectGeoSearch: true,
  connectHierarchicalMenu: true,
  connectHighlight: true,
  connectHits: true,
  connectHitsPerPage: true,
  connectInfiniteHits: true,
  connectMenu: true,
  connectNumericMenu: true,
  connectPagination: true,
  connectPoweredBy: true,
  connectQueryRules: true,
  connectRange: true,
  connectRefinementList: true,
  connectScrollTo: true,
  connectSearchBox: true,
  connectSmartSort: true,
  connectSortBy: true,
  connectStateResults: true,
  connectStats: true,
  connectToggleRefinement: true,
  connectHitInsights: true,
  connectVoiceSearch: true
};
Object.defineProperty(exports, "createConnector", {
  enumerable: true,
  get: function get() {
    return _createConnector.default;
  }
});
Object.defineProperty(exports, "instantSearchContext", {
  enumerable: true,
  get: function get() {
    return _context.instantSearchContext;
  }
});
Object.defineProperty(exports, "InstantSearchConsumer", {
  enumerable: true,
  get: function get() {
    return _context.InstantSearchConsumer;
  }
});
Object.defineProperty(exports, "InstantSearchProvider", {
  enumerable: true,
  get: function get() {
    return _context.InstantSearchProvider;
  }
});
Object.defineProperty(exports, "HIGHLIGHT_TAGS", {
  enumerable: true,
  get: function get() {
    return _highlight.HIGHLIGHT_TAGS;
  }
});
Object.defineProperty(exports, "version", {
  enumerable: true,
  get: function get() {
    return _version.default;
  }
});
Object.defineProperty(exports, "translatable", {
  enumerable: true,
  get: function get() {
    return _translatable.default;
  }
});
Object.defineProperty(exports, "Configure", {
  enumerable: true,
  get: function get() {
    return _Configure.default;
  }
});
Object.defineProperty(exports, "ExperimentalConfigureRelatedItems", {
  enumerable: true,
  get: function get() {
    return _ConfigureRelatedItems.default;
  }
});
Object.defineProperty(exports, "QueryRuleContext", {
  enumerable: true,
  get: function get() {
    return _QueryRuleContext.default;
  }
});
Object.defineProperty(exports, "Index", {
  enumerable: true,
  get: function get() {
    return _Index.default;
  }
});
Object.defineProperty(exports, "InstantSearch", {
  enumerable: true,
  get: function get() {
    return _InstantSearch.default;
  }
});
Object.defineProperty(exports, "connectAutoComplete", {
  enumerable: true,
  get: function get() {
    return _connectAutoComplete.default;
  }
});
Object.defineProperty(exports, "connectBreadcrumb", {
  enumerable: true,
  get: function get() {
    return _connectBreadcrumb.default;
  }
});
Object.defineProperty(exports, "connectConfigure", {
  enumerable: true,
  get: function get() {
    return _connectConfigure.default;
  }
});
Object.defineProperty(exports, "EXPERIMENTAL_connectConfigureRelatedItems", {
  enumerable: true,
  get: function get() {
    return _connectConfigureRelatedItems.default;
  }
});
Object.defineProperty(exports, "connectCurrentRefinements", {
  enumerable: true,
  get: function get() {
    return _connectCurrentRefinements.default;
  }
});
Object.defineProperty(exports, "connectGeoSearch", {
  enumerable: true,
  get: function get() {
    return _connectGeoSearch.default;
  }
});
Object.defineProperty(exports, "connectHierarchicalMenu", {
  enumerable: true,
  get: function get() {
    return _connectHierarchicalMenu.default;
  }
});
Object.defineProperty(exports, "connectHighlight", {
  enumerable: true,
  get: function get() {
    return _connectHighlight.default;
  }
});
Object.defineProperty(exports, "connectHits", {
  enumerable: true,
  get: function get() {
    return _connectHits.default;
  }
});
Object.defineProperty(exports, "connectHitsPerPage", {
  enumerable: true,
  get: function get() {
    return _connectHitsPerPage.default;
  }
});
Object.defineProperty(exports, "connectInfiniteHits", {
  enumerable: true,
  get: function get() {
    return _connectInfiniteHits.default;
  }
});
Object.defineProperty(exports, "connectMenu", {
  enumerable: true,
  get: function get() {
    return _connectMenu.default;
  }
});
Object.defineProperty(exports, "connectNumericMenu", {
  enumerable: true,
  get: function get() {
    return _connectNumericMenu.default;
  }
});
Object.defineProperty(exports, "connectPagination", {
  enumerable: true,
  get: function get() {
    return _connectPagination.default;
  }
});
Object.defineProperty(exports, "connectPoweredBy", {
  enumerable: true,
  get: function get() {
    return _connectPoweredBy.default;
  }
});
Object.defineProperty(exports, "connectQueryRules", {
  enumerable: true,
  get: function get() {
    return _connectQueryRules.default;
  }
});
Object.defineProperty(exports, "connectRange", {
  enumerable: true,
  get: function get() {
    return _connectRange.default;
  }
});
Object.defineProperty(exports, "connectRefinementList", {
  enumerable: true,
  get: function get() {
    return _connectRefinementList.default;
  }
});
Object.defineProperty(exports, "connectScrollTo", {
  enumerable: true,
  get: function get() {
    return _connectScrollTo.default;
  }
});
Object.defineProperty(exports, "connectSearchBox", {
  enumerable: true,
  get: function get() {
    return _connectSearchBox.default;
  }
});
Object.defineProperty(exports, "connectSmartSort", {
  enumerable: true,
  get: function get() {
    return _connectSmartSort.default;
  }
});
Object.defineProperty(exports, "connectSortBy", {
  enumerable: true,
  get: function get() {
    return _connectSortBy.default;
  }
});
Object.defineProperty(exports, "connectStateResults", {
  enumerable: true,
  get: function get() {
    return _connectStateResults.default;
  }
});
Object.defineProperty(exports, "connectStats", {
  enumerable: true,
  get: function get() {
    return _connectStats.default;
  }
});
Object.defineProperty(exports, "connectToggleRefinement", {
  enumerable: true,
  get: function get() {
    return _connectToggleRefinement.default;
  }
});
Object.defineProperty(exports, "connectHitInsights", {
  enumerable: true,
  get: function get() {
    return _connectHitInsights.default;
  }
});
Object.defineProperty(exports, "connectVoiceSearch", {
  enumerable: true,
  get: function get() {
    return _connectVoiceSearch.default;
  }
});

var _createConnector = _interopRequireDefault(require("./core/createConnector"));

var _context = require("./core/context");

var _highlight = require("./core/highlight");

var _version = _interopRequireDefault(require("./core/version"));

var _translatable = _interopRequireDefault(require("./core/translatable"));

var _Configure = _interopRequireDefault(require("./widgets/Configure"));

var _ConfigureRelatedItems = _interopRequireDefault(require("./widgets/ConfigureRelatedItems"));

var _QueryRuleContext = _interopRequireDefault(require("./widgets/QueryRuleContext"));

var _Index = _interopRequireDefault(require("./widgets/Index"));

var _InstantSearch = _interopRequireDefault(require("./widgets/InstantSearch"));

var _connectAutoComplete = _interopRequireDefault(require("./connectors/connectAutoComplete"));

var _connectBreadcrumb = _interopRequireDefault(require("./connectors/connectBreadcrumb"));

var _connectConfigure = _interopRequireDefault(require("./connectors/connectConfigure"));

var _connectConfigureRelatedItems = _interopRequireDefault(require("./connectors/connectConfigureRelatedItems"));

var _connectCurrentRefinements = _interopRequireDefault(require("./connectors/connectCurrentRefinements"));

var _connectGeoSearch = _interopRequireDefault(require("./connectors/connectGeoSearch"));

var _connectHierarchicalMenu = _interopRequireDefault(require("./connectors/connectHierarchicalMenu"));

var _connectHighlight = _interopRequireDefault(require("./connectors/connectHighlight"));

var _connectHits = _interopRequireDefault(require("./connectors/connectHits"));

var _connectHitsPerPage = _interopRequireDefault(require("./connectors/connectHitsPerPage"));

var _connectInfiniteHits = _interopRequireDefault(require("./connectors/connectInfiniteHits"));

var _connectMenu = _interopRequireDefault(require("./connectors/connectMenu"));

var _connectNumericMenu = _interopRequireDefault(require("./connectors/connectNumericMenu"));

var _connectPagination = _interopRequireDefault(require("./connectors/connectPagination"));

var _connectPoweredBy = _interopRequireDefault(require("./connectors/connectPoweredBy"));

var _connectQueryRules = _interopRequireDefault(require("./connectors/connectQueryRules"));

var _connectRange = _interopRequireDefault(require("./connectors/connectRange"));

var _connectRefinementList = _interopRequireDefault(require("./connectors/connectRefinementList"));

var _connectScrollTo = _interopRequireDefault(require("./connectors/connectScrollTo"));

var _connectSearchBox = _interopRequireDefault(require("./connectors/connectSearchBox"));

var _connectSmartSort = _interopRequireDefault(require("./connectors/connectSmartSort"));

var _connectSortBy = _interopRequireDefault(require("./connectors/connectSortBy"));

var _connectStateResults = _interopRequireDefault(require("./connectors/connectStateResults"));

var _connectStats = _interopRequireDefault(require("./connectors/connectStats"));

var _connectToggleRefinement = _interopRequireDefault(require("./connectors/connectToggleRefinement"));

var _connectHitInsights = _interopRequireDefault(require("./connectors/connectHitInsights"));

var _connectVoiceSearch = _interopRequireDefault(require("./connectors/connectVoiceSearch"));

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});
