// popsicle-redirects/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = require("url");
/**
 * Redirection types to handle.
 */
var REDIRECT_TYPE;
(function (REDIRECT_TYPE) {
    REDIRECT_TYPE[REDIRECT_TYPE["FOLLOW_WITH_GET"] = 0] = "FOLLOW_WITH_GET";
    REDIRECT_TYPE[REDIRECT_TYPE["FOLLOW_WITH_CONFIRMATION"] = 1] = "FOLLOW_WITH_CONFIRMATION";
})(REDIRECT_TYPE || (REDIRECT_TYPE = {}));
/**
 * Possible redirection status codes.
 */
const REDIRECT_STATUS = {
    "301": REDIRECT_TYPE.FOLLOW_WITH_GET,
    "302": REDIRECT_TYPE.FOLLOW_WITH_GET,
    "303": REDIRECT_TYPE.FOLLOW_WITH_GET,
    "307": REDIRECT_TYPE.FOLLOW_WITH_CONFIRMATION,
    "308": REDIRECT_TYPE.FOLLOW_WITH_CONFIRMATION
};
/**
 * Maximum redirects error.
 */
class MaxRedirectsError extends Error {
    constructor(request, message) {
        super(message);
        this.request = request;
        this.code = "EMAXREDIRECTS";
    }
}
exports.MaxRedirectsError = MaxRedirectsError;
/**
 * Middleware function for following HTTP redirects.
 */
function redirects(fn, maxRedirects = 5, confirmRedirect = () => false) {
    return async function (initReq, done) {
        let req = initReq.clone();
        let redirectCount = 0;
        while (redirectCount++ < maxRedirects) {
            const res = await fn(req, done);
            const redirect = REDIRECT_STATUS[res.status];
            if (redirect === undefined || !res.headers.has("Location"))
                return res;
            const newUrl = url_1.resolve(req.url, res.headers.get("Location")); // tslint:disable-line
            await res.destroy(); // Ignore the result of the response on redirect.
            req.signal.emit("redirect", newUrl);
            if (redirect === REDIRECT_TYPE.FOLLOW_WITH_GET) {
                const method = initReq.method.toUpperCase() === "HEAD" ? "HEAD" : "GET";
                req = initReq.clone();
                req.url = newUrl;
                req.method = method;
                req.$rawBody = null; // Override internal raw body.
                // No body will be sent with this redirect.
                req.headers.set("Content-Length", "0");
                continue;
            }
            if (redirect === REDIRECT_TYPE.FOLLOW_WITH_CONFIRMATION) {
                const { method } = req;
                // Following HTTP spec by automatically redirecting with GET/HEAD.
                if (method.toUpperCase() === "GET" || method.toUpperCase() === "HEAD") {
                    req = initReq.clone();
                    req.url = newUrl;
                    req.method = method;
                    continue;
                }
                // Allow the user to confirm redirect according to HTTP spec.
                if (confirmRedirect(req, res)) {
                    req = initReq.clone();
                    req.url = newUrl;
                    req.method = method;
                    continue;
                }
            }
            return res;
        }
        throw new MaxRedirectsError(req, `Maximum redirects exceeded: ${maxRedirects}`);
    };
}
exports.redirects = redirects;
